src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,541 @@
+        **relevance_logits**: `torch.FloatTensor`` of shape ``(n_passages, )``
+            Outputs of the QA classifier of the DPRReader that corresponds to the scores of each passage
+            to answer the question, compared to all the other passages.
+
+
+    Examples::
+
+        tokenizer = DPRReader.from_pretrained('dpr-base-uncased')
+        model = DPRModel.from_pretrained('dpr-reader-base')
+        question_and_titles_ids = [
+            torch.tensor(tokenizer("Hello, is my dog cute ?", "Dog cuteness"))
+            ]  # One tensor per passage. It corresponds to the concatenation of the question and the context title.
+        texts_ids = [
+            torch.tensor(tokenizer("Hello, my dog is definitely cute", add_special_tokens=False))
+            ]  # One tensor per passage. It corresponds to the context text in which we're looking for the answer.
+        outputs = model(question_and_titles_ids, texts_ids)
+        start_logits = outputs[1]  # The logits of the start of the span
+        end_logits = outputs[2]  # The logits of the end of the span
+
+    """
+
+    def __init__(self, config: DPRConfig):
+        super().__init__(config)
+        self.config = config
+        self.span_predictor = DPRSpanPredictor(config)
+        self.init_weights()
+
+    def forward(self, question_and_titles_ids: List[Tensor], texts_ids: List[Tensor],) -> Tuple[Tensor, ...]:
+        assert len(question_and_titles_ids) == len(
+            texts_ids
+        ), "There should be as many `question_and_titles_ids` than `texts_ids` but got sizes {}!={}".format(
+            len(question_and_titles_ids), len(texts_ids)
+        )
+        device = question_and_titles_ids[0].device
+        n_contexts = len(question_and_titles_ids)
+        input_ids = torch.ones((n_contexts, self.config.sequence_length), dtype=torch.int64, device=device) * int(
+            self.config.pad_token_id
+        )
+        for i in range(n_contexts):
+            question_and_title_ids = question_and_titles_ids[i]
+            text_ids = texts_ids[i]
+            len_question_and_title = question_and_title_ids.shape[1]
+            _, len_txt = text_ids.size()
+            input_ids[i, 0:len_question_and_title] = question_and_title_ids
+            input_ids[i, len_question_and_title : len_question_and_title + len_txt] = text_ids
+        input_ids = input_ids.unsqueeze(0)
+        attention_mask = input_ids != self.config.pad_token_id
+        attention_mask = attention_mask.to(device=device)
+        start_logits, end_logits, relevance_logits = self.span_predictor(input_ids, attention_mask)
+        return input_ids, start_logits, end_logits, relevance_logits
+
+    def generate(
+        self,
+        question_and_titles_ids: List[Tensor],
+        texts_ids: List[Tensor],
+        num_spans: int = 16,
+        max_answer_length: int = 64,
+        num_spans_per_passage: int = 4,
+    ) -> List[DPRReaderOutput]:
```

Comment:
I'm also wondering if this whole decoding method should not rather be in the tokenizer (like most of the previous forward method).

There is no matrix multiplications or nothing that seems to be accelerated by a DL framework here, it seems to be all about input manipulations (unless I missed something) so we could have it in the tokenizer which would let us better separate the concerns and, for instance, have TF/JAX model with the same code, jit the model (not possible with NamedTuple I think) or use ONNX, etc...

Reply:
I agree. We can imagine something like
```python
start_logits, end_logits, relevance_logits = reader(input_ids)
best_span_ids = tokenizer.extract_best_span(input_ids, start_logits, end_logits, relevance_logits)
answer = tokenizer.decode(best_span_ids)
```
This is linked to the way I'll change the tokenizer of the reader to accept three input strings. I think I'll keep those changes about the tokenizer for the next PR

Reply:
Will the next PR with the tokenizer introduce breaking changes when compared to this PR? Wouldn't it be better to do all the changes at once in this PR?